File: solutions/Leetcode_752/leetcode_752.cc
#include <queue>
#include <string>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
    int openLock(vector<string>& deadends, string target)
    {
        unordered_set<string> visited(deadends.begin(), deadends.end());
        if (visited.find("0000") != visited.end())
            return -1;
        queue<string> q;
        q.push("0000");
        int minimum_turn = -1;
        while (!q.empty())
        {
            int q_len = q.size();
            minimum_turn++;
            for (int i = 0; i < q_len; i++)
            {
                string num = q.front();
                q.pop();
                if (num == target)
                    return minimum_turn;
                if (visited.find(num) != visited.end())
                    continue;
                visited.insert(num);

                for (int i = 0; i < 4; i++)
                {
                    string tmp = num;
                    tmp[i] = (num[i] - '0' + 1) % 10 + '0';
                    q.push(tmp);
                    tmp[i] = (num[i] - '0' - 1 + 10) % 10 + '0';
                    q.push(tmp);
                }
            }
        }
        return -1;
    }
};

File: solutions/Leetcode_752/leetcode_752.py
from collections import deque
from typing import List


class Solution1:
    def open_lock(self, deadends: List[str], target: str) -> int:
        if '0000' in deadends:
            return -1
        visited = set(deadends)
        minimum_turns = -1
        queue = deque(['0000'])
        while queue:
            minimum_turns += 1
            for _ in range(len(queue)):
                num = queue.popleft()
                if num == target:
                    return minimum_turns
                if num in visited:
                    continue
                visited.add(num)
                for i in range(4):
                    up = num[:i] + str((int(num[i]) + 1) % 10) + num[i+1:]
                    down = num[:i] + str((int(num[i]) - 1) % 10) + num[i+1:]
                    queue.append(up)
                    queue.append(down)
        return -1